staging/src/k8s.io/client-go/kubernetes/typed/events/v1/event_expansion.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"fmt"
+
+	eventsv1 "k8s.io/api/events/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// The EventExpansion interface allows manually adding extra methods to the EventInterface.
+// TODO: Add querying functions to the event expansion
+type EventExpansion interface {
+	// CreateWithEventNamespace is the same as a Create
+	// except that it sends the request to the event.Namespace.
+	CreateWithEventNamespace(event *eventsv1.Event) (*eventsv1.Event, error)
+	// UpdateWithEventNamespace is the same as a Update
+	// except that it sends the request to the event.Namespace.
+	UpdateWithEventNamespace(event *eventsv1.Event) (*eventsv1.Event, error)
+	// PatchWithEventNamespace is the same as an Update
+	// except that it sends the request to the event.Namespace.
+	PatchWithEventNamespace(event *eventsv1.Event, data []byte) (*eventsv1.Event, error)
+}
```

Comment:
do these make sense to propagate to the events/v1 client? should we collapse onto the normal clientset methods? it's unclear to me why we want these in the event client.

at the very least, we should make the signatures match the clientset methods (context, options, etc)

Reply:
Thanks for catching that Jordan.
I'm fairly sure that we used to use it more widely now, but either I'm wrong or it has changed. It seems that those methods are really used only in a single place, being event_broadcaster.

So instead of creating the EventExpansion here, lt's simply inline their implemention in broadcaster and get rid of those.

Reply:
something along these lines would work:
```diff
diff --git a/staging/src/k8s.io/client-go/tools/events/event_broadcaster.go b/staging/src/k8s.io/client-go/tools/events/event_broadcaster.go
index ceedc62d762..ecb6a36e12b 100644
--- a/staging/src/k8s.io/client-go/tools/events/event_broadcaster.go
+++ b/staging/src/k8s.io/client-go/tools/events/event_broadcaster.go
@@ -17,6 +17,7 @@ limitations under the License.
 package events
 
 import (
+	"context"
 	"os"
 	"sync"
 	"time"
@@ -26,6 +27,7 @@ import (
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/clock"
 	"k8s.io/apimachinery/pkg/util/json"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
@@ -72,22 +74,31 @@ type eventBroadcasterImpl struct {
 // TODO: this makes it easier for testing purpose and masks the logic of performing API calls.
 // Note that rollbacking to raw clientset should also be transparent.
 type EventSinkImpl struct {
-	Interface typedeventsv1.EventInterface
+	Interface typedeventsv1.EventsV1Interface
 }
 
 // Create is the same as CreateWithEventNamespace of the EventExpansion
 func (e *EventSinkImpl) Create(event *eventsv1.Event) (*eventsv1.Event, error) {
-	return e.Interface.CreateWithEventNamespace(event)
+	// if event.Namespace == "" {
+	// 	... error ...
+	// }
+	return e.Interface.Events(event.Namespace).Create(context.TODO(), event, metav1.CreateOptions{})
 }
 
 // Update is the same as UpdateithEventNamespace of the EventExpansion
 func (e *EventSinkImpl) Update(event *eventsv1.Event) (*eventsv1.Event, error) {
-	return e.Interface.UpdateWithEventNamespace(event)
+	// if event.Namespace == "" {
+	// 	... error ...
+	// }
+	return e.Interface.Events(event.Namespace).Update(context.TODO(), event, metav1.UpdateOptions{})
 }
 
 // Patch is the same as PatchWithEventNamespace of the EventExpansion
 func (e *EventSinkImpl) Patch(event *eventsv1.Event, data []byte) (*eventsv1.Event, error) {
-	return e.Interface.PatchWithEventNamespace(event, data)
+	// if event.Namespace == "" {
+	// 	... error ...
+	// }
+	return e.Interface.Events(event.Namespace).Patch(context.TODO(), event.Name, types.StrategicMergePatchType, data, metav1.PatchOptions{})
 }
 
 // NewBroadcaster Creates a new event broadcaster.
@@ -321,7 +332,7 @@ func (e *eventBroadcasterImpl) StartRecordingToSink(stopCh <-chan struct{}) {
 type eventBroadcasterAdapterImpl struct {
 	coreClient          typedv1core.EventsGetter
 	coreBroadcaster     record.EventBroadcaster
-	eventsv1Client      typedeventsv1.EventsGetter
+	eventsv1Client      typedeventsv1.EventsV1Interface
 	eventsv1Broadcaster EventBroadcaster
 }
 
@@ -331,7 +342,7 @@ func NewEventBroadcasterAdapter(client clientset.Interface) EventBroadcasterAdap
 	eventClient := &eventBroadcasterAdapterImpl{}
 	if _, err := client.Discovery().ServerResourcesForGroupVersion(eventsv1.SchemeGroupVersion.String()); err == nil {
 		eventClient.eventsv1Client = client.EventsV1()
-		eventClient.eventsv1Broadcaster = NewBroadcaster(&EventSinkImpl{Interface: eventClient.eventsv1Client.Events("")})
+		eventClient.eventsv1Broadcaster = NewBroadcaster(&EventSinkImpl{Interface: eventClient.eventsv1Client})
 	}
 	// Even though there can soon exist cases when coreBroadcaster won't really be needed,
 	// we create it unconditionally because its overhead is minor and will simplify using usage
```

and usage changes like this:
```diff
- recorderFactory := profile.NewRecorderFactory(events.NewBroadcaster(&events.EventSinkImpl{Interface: client.EventsV1().Events("")}))
+ recorderFactory := profile.NewRecorderFactory(events.NewBroadcaster(&events.EventSinkImpl{Interface: client.EventsV1()}))
```